src/pages/studyView/StudyViewUtils.tsx

```diff
@@ -2417,3 +2418,102 @@ export function getFilteredAndCompressedDataIntervalFilters(
             : undefined;
     return { start, end } as any;
 }
+
+export async function getClinicalDataBySamples(samples: Sample[]) {
+    let clinicalData: {
+        [sampleId: string]: { [attributeId: string]: string };
+    } = {};
+
+    let sampleClinicalData = await defaultClient.fetchClinicalDataUsingPOST({
+        clinicalDataType: 'SAMPLE',
+        clinicalDataMultiStudyFilter: {
+            identifiers: _.map(samples, sample => {
+                return {
+                    entityId: sample.sampleId,
+                    studyId: sample.studyId,
+                };
+            }),
+        } as ClinicalDataMultiStudyFilter,
+    });
+
+    _.forEach(sampleClinicalData, item => {
+        clinicalData[item.uniqueSampleKey] = {
+            ...(clinicalData[item.uniqueSampleKey] || {}),
+            [item.clinicalAttributeId]: item.value,
+        };
+    });
+
+    let patientClinicalData = await defaultClient.fetchClinicalDataUsingPOST({
+        clinicalDataType: ClinicalDataTypeEnum.PATIENT,
+        clinicalDataMultiStudyFilter: {
+            identifiers: _.map(samples, sample => {
+                return {
+                    entityId: sample.patientId,
+                    studyId: sample.studyId,
+                };
+            }),
+        } as ClinicalDataMultiStudyFilter,
+    });
+
+    const patientSamplesMap = _.groupBy(
+        samples,
+        sample => sample.uniquePatientKey
+    );
+
+    _.forEach(patientClinicalData, item => {
+        (patientSamplesMap[item.uniquePatientKey] || []).forEach(sample => {
+            clinicalData[sample.uniqueSampleKey] = {
+                ...(clinicalData[sample.uniqueSampleKey] || {}),
+                [item.clinicalAttributeId]: item.value,
+            };
+        });
+    });
+    return clinicalData;
+}
+
+export function updateChartIds(
+    chartSettings: ChartUserSetting[]
+): ChartUserSetting[] {
+    const customChartRegex = /^CUSTOM_FILTERS_\d+/i;
+    const chartIdWithDataTypeRegex = /^(?:PATIENT_|SAMPLE_)([a-zA-Z0-9_]+)/i;
+    let numberOfClinicalAttributeCharts = 0;
+    let numberOfChartRequiringUpdates = 0;
+    const specialChartKeySet = _.reduce(
+        UniqueKey,
+        (acc, next) => {
+            acc[next] = true;
+            return acc;
+        },
+        {} as { [id: string]: boolean }
+    );
+
+    chartSettings.forEach(chartSetting => {
+        let customChartmatch = chartSetting.id.match(customChartRegex);
+        if (
+            !customChartmatch &&
+            specialChartKeySet[chartSetting.id] === undefined
+        ) {
+            numberOfClinicalAttributeCharts++;
+            let match = chartSetting.id.match(chartIdWithDataTypeRegex);
+            if (!!match) {
+                numberOfChartRequiringUpdates++;
+            }
+        }
+    });
+
+    // Only update ids if all the chart ids require updates.
```

Comment:
this comment could use further elaboration.  i just don't really understand business logic here.